File: android/app/src/main/java/de/kjell/zencompanion/data/SearchEngine.kt
package de.kjell.zencompanion.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

/**
 * A search provider: a display name plus a URL template whose `{query}`
 * placeholder is replaced with the URL-encoded query.
 *
 * Built-ins live in [SearchEngines.builtIn]; user-defined engines are stored
 * on the device only and never synced. Mirrors the iOS `SearchEngine`.
 */
data class SearchEngine(
    val id: String,
    val displayName: String,
    val template: String,
    val isBuiltIn: Boolean,
) {
    /** Replaces `{query}` in the template with the UTF-8 URL-encoded query. */
    fun formatQuery(query: String): String {
        val encoded = URLEncoder.encode(query, StandardCharsets.UTF_8.name())
        return template.replace(SearchEngines.PLACEHOLDER, encoded)
    }

    fun searchURL(query: String): String = formatQuery(query)

    companion object {
        val DUCKDUCKGO = SearchEngine("duckduckgo", "DuckDuckGo", "https://duckduckgo.com/?q={query}", true)
        val GOOGLE = SearchEngine("google", "Google", "https://www.google.com/search?q={query}", true)
        val ECOSIA = SearchEngine("ecosia", "Ecosia", "https://www.ecosia.org/search?q={query}", true)
        val BRAVE = SearchEngine("brave", "Brave", "https://search.brave.com/search?q={query}", true)
        val BING = SearchEngine("bing", "Bing", "https://www.bing.com/search?q={query}", true)
    }
}

/** Why a custom-engine draft cannot be saved yet. */
enum class SearchEngineValidation {
    EMPTY_NAME,
    INVALID_URL,
    MISSING_PLACEHOLDER,
}

/**
 * Validation and the "paste a search link" template derivation for custom
 * engines. Pure and side-effect free, so it is unit-testable on the JVM.
 */
object SearchEngineTemplate {
    const val PLACEHOLDER = "{query}"

    /** null when the draft is ready to save, otherwise the first problem. */
    fun validate(name: String, template: String): SearchEngineValidation? {
        if (name.isBlank()) return SearchEngineValidation.EMPTY_NAME
        val trimmed = template.trim()
        val lower = trimmed.lowercase()
        if (!lower.startsWith("http://") && !lower.startsWith("https://")) {
            return SearchEngineValidation.INVALID_URL
        }
        if (runCatching { URI(trimmed.replace(PLACEHOLDER, "test")) }.isFailure) {
            return SearchEngineValidation.INVALID_URL
        }
        if (!trimmed.contains(PLACEHOLDER)) return SearchEngineValidation.MISSING_PLACEHOLDER
        return null
    }

    /**
     * Turns a real search URL (e.g. copied from the address bar) into a
     * template by replacing the query parameter's value with `{query}`.
     * Prefers well-known parameter names (`q`, `query`, …), else the last
     * parameter that has a value.
     */
    fun derive(fromPastedURL: String): String? {
        val trimmed = fromPastedURL.trim()
        val lower = trimmed.lowercase()
        if (!lower.startsWith("http://") && !lower.startsWith("https://")) return null
        val question = trimmed.indexOf('?')
        if (question < 0) return null

        val base = trimmed.substring(0, question)
        var remainder = trimmed.substring(question + 1)
        var fragment = ""
        val hash = remainder.indexOf('#')
        if (hash >= 0) {
            fragment = remainder.substring(hash)
            remainder = remainder.substring(0, hash)
        }
        if (remainder.isEmpty()) return null

        val preferred = setOf("q", "query", "search", "p", "text", "wd", "k")
        val pairs = remainder.split('&').toMutableList()

        fun keyOf(pair: String): String {
            val rawKey = pair.substringBefore('=')
            return runCatching { URLDecoder.decode(rawKey, "UTF-8") }.getOrDefault(rawKey).lowercase()
        }

        fun valueOf(pair: String): String {
            val index = pair.indexOf('=')
            return if (index < 0) "" else pair.substring(index + 1)
        }

        var chosen: Int? = null
        for ((index, pair) in pairs.withIndex()) {
            if (valueOf(pair).isEmpty()) continue
            if (preferred.contains(keyOf(pair))) {
                chosen = index
                break
            }
        }
        if (chosen == null) {
            chosen = pairs.indexOfLast { valueOf(it).isNotEmpty() }.takeIf { it >= 0 }
        }
        val index = chosen ?: return null

        val originalKey = pairs[index].substringBefore('=')
        pairs[index] = "$originalKey=$PLACEHOLDER"
        return base + "?" + pairs.joinToString("&") + fragment
    }

    /**
     * Best-effort display name for a pasted search link: the registrable
     * domain's label, so `de.search.yahoo.com` -> "Yahoo" and `www.bbc.co.uk`
     * -> "Bbc" (not the sub-domain "de" or the "www").
     */
    fun suggestedName(fromTemplate: String): String? {
        val host = runCatching { URI(fromTemplate.substringBefore('?')).host }.getOrNull() ?: return null
        val labels = host.lowercase().split('.').filter { it.isNotEmpty() }.toMutableList()
        if (labels.firstOrNull() == "www") labels.removeAt(0)
        if (labels.isEmpty()) return null

        val label = when {
            labels.size == 1 -> labels[0]
            labels.size >= 3 && compoundSuffixes.contains(labels.takeLast(2).joinToString(".")) ->
                labels[labels.size - 3]
            else -> labels[labels.size - 2]
        }
        return label.replaceFirstChar { it.uppercase() }
    }

    /**
     * Common country-code second-level domains, so `co.uk` style suffixes are
     * skipped when looking for the registrable label.
     */
    private val compoundSuffixes = setOf(
        "co.uk", "org.uk", "ac.uk", "gov.uk", "me.uk", "net.uk", "sch.uk",
        "com.au", "net.au", "org.au", "edu.au", "gov.au",
        "co.nz", "net.nz", "org.nz", "govt.nz",
        "co.jp", "or.jp", "ne.jp", "ac.jp", "go.jp",
        "com.br", "net.br", "org.br", "gov.br",
        "co.in", "net.in", "org.in", "gov.in",
        "com.cn", "net.cn", "org.cn", "gov.cn",
        "co.za", "org.za", "gov.za",
        "co.kr", "or.kr",
        "com.mx", "com.tr", "com.sg", "com.hk", "com.tw", "co.id", "co.il",
    )
}

/**
 * The engine catalogue and persistence. The selection is stored as the engine
 * id (built-in or custom), so a stored value survives and unknown ids fall
 * back to DuckDuckGo — including after a selected custom engine is deleted.
 */
object SearchEngines {
    const val PLACEHOLDER = "{query}"

    const val PREFS_KEY = "zen_companion_search_engine"
    const val LEGACY_PREFS_KEY = "selected_search_engine"
    const val CUSTOM_PREFS_KEY = "zen_companion_custom_search_engines"
    const val PREFS_NAME = "zen_prefs"

    val builtIn: List<SearchEngine> = listOf(
        SearchEngine.DUCKDUCKGO,
        SearchEngine.GOOGLE,
        SearchEngine.ECOSIA,
        SearchEngine.BRAVE,
        SearchEngine.BING,
    )

    @Volatile
    private var cachedEngine: SearchEngine? = null

    fun getPreferences(context: Context) =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /** Built-ins in stable order, then the user's own engines. */
    fun all(context: Context): List<SearchEngine> = builtIn + custom(context)

    /** User-defined engines, in the order they were added. */
    fun custom(context: Context): List<SearchEngine> =
        decodeCustom(getPreferences(context).getString(CUSTOM_PREFS_KEY, null))

    fun setCustom(context: Context, engines: List<SearchEngine>) {
        getPreferences(context).edit()
            .putString(CUSTOM_PREFS_KEY, encodeCustom(engines.filter { !it.isBuiltIn }))
            .apply()
    }

    /** Resolves the persisted selection against [candidates]. */
    fun resolve(context: Context): SearchEngine {
        val raw = getPreferences(context).getString(PREFS_KEY, null)
            ?: getPreferences(context).getString(LEGACY_PREFS_KEY, null)
        return fromKey(raw, all(context))
    }

    fun set(context: Context, engine: SearchEngine) {
        cachedEngine = engine
        getPreferences(context).edit()
            .putString(PREFS_KEY, engine.id)
            .putString(LEGACY_PREFS_KEY, engine.id)
            .apply()
    }

    /** Matches an id or display name (case-insensitive); defaults to DuckDuckGo. */
    fun fromKey(key: String?, candidates: List<SearchEngine> = builtIn): SearchEngine {
        if (key.isNullOrBlank()) return SearchEngine.DUCKDUCKGO
        return candidates.firstOrNull {
            it.id.equals(key, ignoreCase = true) || it.displayName.equals(key, ignoreCase = true)
        } ?: SearchEngine.DUCKDUCKGO
    }

    /** Cached selection for the address bar; requires [AppContextHolder]. */
    var current: SearchEngine
        get() {
            cachedEngine?.let { return it }
            val ctx = runCatching { AppContextHolder.appContext }.getOrNull()
            return if (ctx != null) {
                val engine = resolve(ctx)
                cachedEngine = engine
                engine
            } else {
                SearchEngine.DUCKDUCKGO
            }
        }
        set(value) {
            cachedEngine = value
            val ctx = runCatching { AppContextHolder.appContext }.getOrNull()
            if (ctx != null) set(ctx, value)
        }

    fun invalidateCache() {
        cachedEngine = null
    }

    private fun encodeCustom(engines: List<SearchEngine>): String {
        val array = JSONArray()
        engines.forEach { engine ->
            array.put(
                JSONObject().apply {
                    put("id", engine.id)
                    put("name", engine.displayName)
                    put("template", engine.template)
                },
            )
        }
        return array.toString()
    }

    private fun decodeCustom(raw: String?): List<SearchEngine> {
        if (raw.isNullOrBlank()) return emptyList()
        return runCatching {
            val array = JSONArray(raw)
            (0 until array.length()).mapNotNull { index ->
                val obj = array.optJSONObject(index) ?: return@mapNotNull null
                val id = obj.optString("id").takeIf { it.isNotBlank() } ?: return@mapNotNull null
                val template = obj.optString("template").takeIf { it.isNotBlank() } ?: return@mapNotNull null
                SearchEngine(
                    id = id,
                    displayName = obj.optString("name").ifBlank { "Search" },
                    template = template,
                    isBuiltIn = false,
                )
            }
        }.getOrDefault(emptyList())
    }
}

File: ios/Shared/Data/SearchEngine.swift
import Foundation

/// A search provider: a display name plus a URL template whose `{query}`
/// placeholder is replaced with the percent-encoded query.
///
/// The built-in engines are the fixed list in `SearchEngines.builtIn`. Users
/// may add their own; those live only on the device and are never synced.
public struct SearchEngine: Identifiable, Hashable, Codable, Sendable {
    public let id: String
    public let displayName: String
    public let template: String
    public let isBuiltIn: Bool

    /// Marker in `template` that receives the encoded query.
    public static let queryPlaceholder = "{query}"

    public init(id: String, displayName: String, template: String, isBuiltIn: Bool) {
        self.id = id
        self.displayName = displayName
        self.template = template
        self.isBuiltIn = isBuiltIn
    }

    /// The search URL for `query`, or nil when the template is unusable.
    public func searchURL(for query: String) -> URL? {
        guard let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else { return nil }
        return URL(string: template.replacingOccurrences(of: SearchEngine.queryPlaceholder, with: encoded))
    }
}

// MARK: - Built-ins

public extension SearchEngine {
    static let duckDuckGo = SearchEngine(
        id: "duckduckgo", displayName: "DuckDuckGo",
        template: "https://duckduckgo.com/?q={query}", isBuiltIn: true
    )
    static let google = SearchEngine(
        id: "google", displayName: "Google",
        template: "https://www.google.com/search?q={query}", isBuiltIn: true
    )
    static let ecosia = SearchEngine(
        id: "ecosia", displayName: "Ecosia",
        template: "https://www.ecosia.org/search?q={query}", isBuiltIn: true
    )
    static let brave = SearchEngine(
        id: "brave", displayName: "Brave",
        template: "https://search.brave.com/search?q={query}", isBuiltIn: true
    )
    static let bing = SearchEngine(
        id: "bing", displayName: "Bing",
        template: "https://www.bing.com/search?q={query}", isBuiltIn: true
    )
}

// MARK: - Catalogue + persistence

/// The engine catalogue and the persisted selection. Reads the App Group
/// first, then standard defaults, and writes to both — mirroring the previous
/// enum so the Share Extension sees the same choice.
public enum SearchEngines {
    public static let builtIn: [SearchEngine] = [.duckDuckGo, .google, .ecosia, .brave, .bing]

    private static let selectedKey = "selected_search_engine"
    private static let customKey = "custom_search_engines"

    /// User-defined engines, in the order they were added.
    public static var custom: [SearchEngine] {
        get {
            guard let data = object(forKey: customKey) as? Data,
                  let decoded = try? JSONDecoder().decode([SearchEngine].self, from: data)
            else { return [] }
            return decoded.filter { !$0.isBuiltIn }
        }
        set {
            let sanitized = newValue.filter { !$0.isBuiltIn }
            guard let data = try? JSONEncoder().encode(sanitized) else { return }
            setObject(data, forKey: customKey)
        }
    }

    /// Built-ins in stable order, then the user's own engines.
    public static var all: [SearchEngine] { builtIn + custom }

    /// The selected engine, resolved by id. Unknown or missing ids fall back
    /// to DuckDuckGo — including after a selected custom engine is deleted.
    public static var current: SearchEngine {
        get {
            guard let id = object(forKey: selectedKey) as? String else { return .duckDuckGo }
            return all.first { $0.id == id } ?? .duckDuckGo
        }
        set { setObject(newValue.id, forKey: selectedKey) }
    }

    /// Appends a custom engine and returns it (caller decides on selection).
    @discardableResult
    public static func addCustom(name: String, template: String) -> SearchEngine {
        let engine = SearchEngine(
            id: UUID().uuidString,
            displayName: name.trimmingCharacters(in: .whitespacesAndNewlines),
            template: template.trimmingCharacters(in: .whitespacesAndNewlines),
            isBuiltIn: false
        )
        custom.append(engine)
        return engine
    }

    /// Removes a custom engine; built-ins are ignored.
    public static func deleteCustom(id: String) {
        custom = custom.filter { $0.id != id }
    }

    // MARK: Storage

    private static func object(forKey key: String) -> Any? {
        AppGroup.defaults.object(forKey: key) ?? UserDefaults.standard.object(forKey: key)
    }

    private static func setObject(_ value: Any?, forKey key: String) {
        AppGroup.defaults.set(value, forKey: key)
        UserDefaults.standard.set(value, forKey: key)
    }
}

// MARK: - Draft validation + paste derivation

/// Why a custom-engine draft cannot be saved yet.
public enum SearchEngineValidation: Equatable {
    case emptyName
    case invalidURL
    case missingPlaceholder
}

/// Validation and the "paste a search link" template derivation for custom
/// engines. Pure and side-effect free, so it is unit-testable.
public enum SearchEngineTemplate {
    public static let placeholder = SearchEngine.queryPlaceholder

    /// nil when the draft is ready to save, otherwise the first problem.
    public static func validate(name: String, template: String) -> SearchEngineValidation? {
        if name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return .emptyName }
        let trimmed = template.trimmingCharacters(in: .whitespacesAndNewlines)
        let lower = trimmed.lowercased()
        guard lower.hasPrefix("http://") || lower.hasPrefix("https://") else { return .invalidURL }
        guard URL(string: trimmed.replacingOccurrences(of: placeholder, with: "test")) != nil else { return .invalidURL }
        guard trimmed.contains(placeholder) else { return .missingPlaceholder }
        return nil
    }

    /// Turns a real search URL (e.g. copied from the address bar) into a
    /// template by replacing the query parameter's value with `{query}`.
    /// Prefers well-known parameter names (`q`, `query`, …), else the last
    /// parameter that has a value.
    public static func derive(fromPastedURL raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        let lower = trimmed.lowercased()
        guard lower.hasPrefix("http://") || lower.hasPrefix("https://") else { return nil }
        guard let questionMark = trimmed.firstIndex(of: "?") else { return nil }

        let base = String(trimmed[..<questionMark])
        var remainder = String(trimmed[trimmed.index(after: questionMark)...])
        var fragment = ""
        if let hash = remainder.firstIndex(of: "#") {
            fragment = String(remainder[hash...])
            remainder = String(remainder[..<hash])
        }
        guard !remainder.isEmpty else { return nil }

        let preferredKeys: Set<String> = ["q", "query", "search", "p", "text", "wd", "k"]
        var pairs = remainder.components(separatedBy: "&")

        func key(of pair: String) -> String {
            let rawKey = pair.components(separatedBy: "=").first ?? pair
            return (rawKey.removingPercentEncoding ?? rawKey).lowercased()
        }
        func value(of pair: String) -> String {
            let parts = pair.components(separatedBy: "=")
            return parts.count > 1 ? parts.dropFirst().joined(separator: "=") : ""
        }

        var chosen: Int?
        for (index, pair) in pairs.enumerated() where !value(of: pair).isEmpty {
            if preferredKeys.contains(key(of: pair)) {
                chosen = index
                break
            }
        }
        if chosen == nil {
            chosen = pairs.lastIndex(where: { !value(of: $0).isEmpty })
        }
        guard let index = chosen else { return nil }

        let originalKey = pairs[index].components(separatedBy: "=").first ?? ""
        pairs[index] = "\(originalKey)=\(placeholder)"
        return base + "?" + pairs.joined(separator: "&") + fragment
    }

    /// Best-effort display name for a pasted search link: the registrable
    /// domain's label, so `de.search.yahoo.com` → "Yahoo" and
    /// `www.bbc.co.uk` → "Bbc" (not the sub-domain "de" or the "www").
    public static func suggestedName(fromTemplate template: String) -> String? {
        guard let base = template.components(separatedBy: "?").first,
              let host = URL(string: base)?.host else { return nil }
        return suggestedName(fromHost: host)
    }

    /// Common country-code second-level domains, so `co.uk` style suffixes are
    /// skipped when looking for the registrable label.
    private static let compoundSuffixes: Set<String> = [
        "co.uk", "org.uk", "ac.uk", "gov.uk", "me.uk", "net.uk", "sch.uk",
        "com.au", "net.au", "org.au", "edu.au", "gov.au",
        "co.nz", "net.nz", "org.nz", "govt.nz",
        "co.jp", "or.jp", "ne.jp", "ac.jp", "go.jp",
        "com.br", "net.br", "org.br", "gov.br",
        "co.in", "net.in", "org.in", "gov.in",
        "com.cn", "net.cn", "org.cn", "gov.cn",
        "co.za", "org.za", "gov.za",
        "co.kr", "or.kr",
        "com.mx", "com.tr", "com.sg", "com.hk", "com.tw", "co.id", "co.il",
    ]

    /// `www.` is stripped; otherwise the label before the public suffix wins.
    static func suggestedName(fromHost host: String) -> String? {
        var labels = host.lowercased().split(separator: ".").map(String.init)
        if labels.first == "www" { labels.removeFirst() }
        guard let first = labels.first else { return nil }

        let label: String
        if labels.count == 1 {
            label = first
        } else if labels.count >= 3,
                  compoundSuffixes.contains(labels.suffix(2).joined(separator: ".")) {
            label = labels[labels.count - 3]
        } else {
            label = labels[labels.count - 2]
        }

        guard !label.isEmpty else { return nil }
        return label.prefix(1).uppercased() + label.dropFirst()
    }
}
